.github/workflows/ci.yml

```diff
@@ -10,6 +10,106 @@ on:
+        uses: actions/cache@v2
+        id: ossl-cache
+        with:
+          path: ${{ github.workspace }}/osslcache
+          key: ${{ matrix.PYTHON.OPENSSL.TYPE }}-${{ matrix.PYTHON.OPENSSL.VERSION }}-${{ env.CONFIG_HASH }}
+        if: matrix.PYTHON.OPENSSL
+      - name: Build custom OpenSSL/LibreSSL
+        run: |
+          shlib_sed() {
+              # modify the shlib version to a unique one to make sure the dynamic
+              # linker doesn't load the system one.
+              sed -i "s/^SHLIB_MAJOR=.*/SHLIB_MAJOR=100/" Makefile
+              sed -i "s/^SHLIB_MINOR=.*/SHLIB_MINOR=0.0/" Makefile
+              sed -i "s/^SHLIB_VERSION_NUMBER=.*/SHLIB_VERSION_NUMBER=100.0.0/" Makefile
+          }
+
+          OPENSSL_DIR="${{ github.workspace }}/osslcache/${TYPE}-${VERSION}-${CONFIG_HASH}"
+          if [[ "${TYPE}" == "openssl" ]]; then
+              curl -O "https://www.openssl.org/source/openssl-${VERSION}.tar.gz"
+              tar zxf "openssl-${VERSION}.tar.gz"
+              pushd "openssl-${VERSION}"
+              # CONFIG_FLAGS comes from the previous step. It's set as global.
+              ./config $CONFIG_FLAGS -fPIC --prefix="${OPENSSL_DIR}"
+              shlib_sed
+              make depend
+              make -j"$(nproc)"
+              # avoid installing the docs on versions of OpenSSL that aren't ancient.
+              # https://github.com/openssl/openssl/issues/6685#issuecomment-403838728
+              make install_sw install_ssldirs
+              popd
+          elif [[ "${TYPE}" == "libressl" ]]; then
+              curl -O "https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-${VERSION}.tar.gz"
+              tar zxf "libressl-${VERSION}.tar.gz"
+              pushd "libressl-${VERSION}"
+              ./config -Wl -Wl,-Bsymbolic-functions -fPIC shared --prefix="${OPENSSL_DIR}"
+              shlib_sed
+              make -j"$(nproc)" install
+              popd
+          fi
+        env:
+          TYPE: ${{ matrix.PYTHON.OPENSSL.TYPE }}
+          VERSION: ${{ matrix.PYTHON.OPENSSL.VERSION }}
+        if: matrix.PYTHON.OPENSSL && steps.ossl-cache.outputs.cache-hit != 'true'
+      - name: Set CFLAGS/LDFLAGS
+        run: |
+          echo "CFLAGS=${CFLAGS} -I${OSSL_PATH}/include" >> $GITHUB_ENV
+          echo "LDFLAGS=${LDFLAGS} -L${OSSL_PATH}/lib -Wl,-rpath=${OSSL_PATH}/lib" >> $GITHUB_ENV
+        if: matrix.PYTHON.OPENSSL
+      - name: Tests
+        run: |
+            echo "Expected OpenSSL path: ${OSSL_PATH}"
+            tox -r --  --color=yes --wycheproof-root=wycheproof
+        env:
+          TOXENV: ${{ matrix.PYTHON.TOXENV }}
+      - name: Upload coverage
+        run: |
+          curl -o codecov.sh -f https://codecov.io/bash || curl -o codecov.sh -f https://codecov.io/bash || curl -o codecov.sh -f https://codecov.io/bash
+          bash codecov.sh -n "tox -e ${{ matrix.PYTHON.TOXENV }} ${{ matrix.PYTHON.OPENSSL }}"
+        if: matrix.PYTHON.COVERAGE == 'true'
```

Comment:
Can we make this `if != "false"` so we don't need the key in every single element?